### PR TITLE
feat: Refactor GDPR stuff into a more readable consent class

### PR DESCRIPTION
**Pull request by benjackwhite**
## Changes

The "GDPR" opt in/out code was reaaaaaallly messy. Been wanting to refactor it for a while but didn't have a strong need until recently (other PR where I was potentially modifying some bits).

* Creates a new `ConsentManager` class to encapsulate the logic
* Removes the spaghetti of gdpr functions from before
* _Fairly_ sure it works as before and hopefully enough tests to prove it
* Removes a few weird code paths such as the localStorage+cookie (which never would have got stored in the cookie)
* Changes the params given to the opt in functions so you can't choose the persistence type - should now always be done via the config and it is now deterministic

## Checklist
- [x] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
- [ ] Accounted for the impact of any changes across different browsers
- [ ] Accounted for backwards compatibility of any changes (no breaking changes in posthog-js!)

### CODE REVIEW

File: src/__tests__/consent.test.ts [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+                it(`should capture an event recording the opt-in action`, () => {
+                    const onCapture = jest.fn()
+                    posthog.on('eventCaptured', onCapture)
+
+                    posthog.opt_in_capturing()
+                    expect(onCapture).toHaveBeenCalledWith(expect.objectContaining({ event: '$opt_in' }))
+
+                    onCapture.mockClear()
+                    const captureEventName = `єνєηт`
+                    const captureProperties = { '𝖕𝖗𝖔𝖕𝖊𝖗𝖙𝖞': `𝓿𝓪𝓵𝓾𝓮` }
```
Comment (pauldambra) [inline src/__tests__/consent.test.ts:147]: wat 🤣

File: src/__tests__/extensions/replay/sessionrecording.test.ts [hunk 2]
```diff
@@ -160,6 +160,7 @@ describe('SessionRecording', () => {
             sessionManager: sessionManager,
             requestRouter: new RequestRouter({ config } as any),
             _addCaptureHook: jest.fn(),
+            consent: { isOptedOut: jest.fn().mockReturnValue(false) },
```
Comment (pauldambra) [inline src/__tests__/extensions/replay/sessionrecording.test.ts:163]:
```suggestion
            consent: { isOptedOut: () => false },
```

is that equivalent, it's not referenced anywhere else in the file 🤷

File: src/consent.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    private get storageKey() {
+        const { token, opt_out_capturing_cookie_prefix } = this.instance.config
+        return (opt_out_capturing_cookie_prefix || OPT_OUT_PREFIX) + token
+    }
+
+    private get storedConsent(): ConsentStatus {
+        const value = this.storage.get(this.storageKey)
+        return value === '1' ? ConsentStatus.GRANTED : value === '0' ? ConsentStatus.DENIED : ConsentStatus.PENDING
```
Comment (pauldambra) [inline src/consent.ts:67]:
```suggestion
        return ConsentStatus[value] || ConsentStatus.PENDING
```

I _think_ that works

Comment (pauldambra) [reply src/consent.ts:67]: arguable that your more explicit version is more clear though so 🤷

**Review by pauldambra — APPROVED** [review]
This looks great!

I messed around with the cypress tests for this a few months ago so I'm relatively confident that they cover this well (although I was focussed on replay consent)

File: src/__tests__/consent.test.ts (hunk 1 above)
Comment (benjackwhite) [reply src/__tests__/consent.test.ts:147]: yeah thats what was there before 😅 seemed fun enough to keep in

File: src/consent.ts (hunk 3 above)
Comment (benjackwhite) [reply src/consent.ts:67]: It's also _not_ the same I think. The stored value is a string so we would have to parse it to a number

Comment (pauldambra) [reply src/consent.ts:67]: Ah, yeah, I missed that... make the enum values strings 🙈 

:)

**Comment by grahamc** [comment]
Hey folks, I just wanted to raise awareness to the folks involved in this PR that I think it broke the SDK and has exposed users who were GDRP compliant, to being non-compliant with GDPR: https://github.com/PostHog/posthog-js/issues/1547

**Pull request merged**